VSCode extension + chruby custom paths via RUBIES = cannot find installation directory for Ruby

### Description

### Reproduction steps
1. Start VSCode in a ruby project dir. Error notification pops up.
<img width="456" alt="Screenshot 2024-04-24 at 2 01 32 PM" src="https://github.com/Shopify/ruby-lsp/assets/157270/4e21cf5a-9cfa-413a-8ac2-da3b048b763d">

```
Failed to activate [object Object] environment: Cannot find installation directory for Ruby version 3.3.1 or ruby-3.3.1
```

The only output that I see in the VSCode output tab for Ruby LSP is:
```
2024-04-24 14:10:27.958 [info] (my-app) Discovered Ruby version 3.3.1 from file:///Users/swrobel/Code/my-app/.ruby-version
```

I don't see an activation command that I can attempt to replicate, as suggested in the [troubleshooting steps](https://github.com/Shopify/ruby-lsp/blob/main/TROUBLESHOOTING.md#try-to-run-the-ruby-activation-manually), but running `chruby 3.3.1` in the VSCode terminal works without error.

### Versions / Environment
* VSCode 1.88.1
* Ruby LSP VSCode version v0.5.19
* Ruby LSP Gem version unknown, but I've run the `Ruby LSP: Update language server gem` command

**VSCode settings.json**
```json
  "rubyLsp.rubyVersionManager": {
    "identifier": "chruby"
  },
```

**.ruby-version file contents:**
```
3.3.1
```

A perhaps unique aspect of my setup is that I install rubies via Homebrew, and make them available in chruby by setting the `RUBIES` env var in my `.zprofile`. This seems to be picked up properly by VSCode, regardless of whether it's launched from the dock icon or a terminal command.
```
$ echo $RUBIES
/opt/homebrew/Cellar/ruby/3.3.1 /opt/homebrew/Cellar/ruby@3.2/3.2.4
```

### Another bug?
As an aside, this extension produces a confusing/inaccurate error message if using a .ruby-version file that doesn't specify a patch version, even though that should be acceptable for `chruby`, ex:
```
3.3
```

The error shown is:
```
Failed to activate [object Object] environment: No .ruby-version file was found
```

@vinistock, I tested the latest, and noticed that the behavior of the new config doesn't exactly match [the description](https://github.com/Shopify/ruby-lsp/pull/1976/files#diff-99d7a3c2c2975fcaf8184c4639922f5ef6c915610b6f8e15ffa8fb8b329828d8R258) provided, since:

```
$ echo $RUBIES
/opt/homebrew/Cellar/ruby/3.3.1 /opt/homebrew/Cellar/ruby@3.2/3.2.4
```
But I needed to set the following in my `settings.json`, omitting the version numbers:

```json
"chrubyRubies": ["/opt/homebrew/Cellar/ruby","/opt/homebrew/Cellar/ruby@3.2"]
```

After adjusting that, it seems like an improvement, but it's now generating a different set of errors:
```
Failed to setup the bundle: Command failed: gem install ruby-lsp
ERROR:  Error installing ruby-lsp:
	ERROR: Failed to build gem native extension.

    current directory: /Users/swrobel/.gem/ruby/3.3.1/gems/prism-0.27.0/ext/prism
/opt/homebrew/opt/ruby/bin/ruby extconf.rb
checking for prism.h in /Users/swrobel/.gem/ruby/3.3.1/gems/prism-0.27.0/include... no
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include=${opt-dir}/include
	--without-opt-include
	--with-opt-lib=${opt-dir}/lib
	--without-opt-lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/homebrew/Cellar/ruby/3.3.1/bin/$(RUBY_BASE_NAME)
extconf.rb:94:in `<main>': prism.h is required (RuntimeError)

extconf failed, exit code 1

Gem files will remain installed in /Users/swrobel/.gem/ruby/3.3.1/gems/prism-0.27.0 for inspection.
Results logged to /Users/swrobel/.gem/ruby/3.3.1/extensions/arm64-darwin-23/3.3.0/prism-0.27.0/gem_make.out
.         See [Troubleshooting](https://github.com/Shopify/ruby-lsp/blob/main/TROUBLESHOOTING.md) for help
```

```
Failed to setup the bundle: Command failed: gem install ruby-lsp
ERROR:  Error installing ruby-lsp:
	ERROR: Failed to build gem native extension.

    current directory: /Users/swrobel/.gem/ruby/3.3.1/gems/prism-0.27.0/ext/prism
/opt/homebrew/opt/ruby/bin/ruby extconf.rb
checking for prism.h in /Users/swrobel/.gem/ruby/3.3.1/gems/prism-0.27.0/include... yes
checking for prism/extension.h in /Users/swrobel/.gem/ruby/3.3.1/gems/prism-0.27.0/ext... yes
checking for whether -fvisibility=hidden is accepted as CFLAGS... yes
Running make build/libprism.a with {"SOEXT"=>"dylib", "AR"=>"/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ar", "ARFLAGS"=>"rcu ", "MAKEDIRS"=>"mkdir -p", "RMALL"=>"rm -fr"}
make: getcwd: No such file or directory
make: *** No rule to make target `build/libprism.a'.  Stop.
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include=${opt-dir}/include
	--without-opt-include
	--with-opt-lib=${opt-dir}/lib
	--without-opt-lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/homebrew/Cellar/ruby/3.3.1/bin/$(RUBY_BASE_NAME)
	--enable-build-debug
	--disable-build-debug
	--enable-build-minimal
	--disable-build-minimal
extconf.rb:69:in `system': Command failed with exit 2: make (RuntimeError)
	from extconf.rb:69:in `block in make'
	from extconf.rb:68:in `chdir'
	from extconf.rb:68:in `make'
	from extconf.rb:132:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /Users/swrobel/.gem/ruby/3.3.1/extensions/arm64-darwin-23/3.3.0/prism-0.27.0/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /Users/swrobel/.gem/ruby/3.3.1/gems/prism-0.27.0 for inspection.
Results logged to /Users/swrobel/.gem/ruby/3.3.1/extensions/arm64-darwin-23/3.3.0/prism-0.27.0/gem_make.out
.         See [Troubleshooting](https://github.com/Shopify/ruby-lsp/blob/main/TROUBLESHOOTING.md) for help
```

--- a/vscode/src/ruby/chruby.ts
+++ b/vscode/src/ruby/chruby.ts
@@ -43,7 +43,7 @@ export class Chruby extends VersionManager {
   async activate(): Promise<ActivationResult> {
     const versionInfo = await this.discoverRubyVersion();
     const rubyUri = await this.findRubyUri(versionInfo);
-    const { defaultGems, gemHome, yjit } =
+    const { defaultGems, gemHome, yjit, version } =
       await this.runActivationScript(rubyUri);
 
     this.outputChannel.info(
@@ -62,12 +62,51 @@ export class Chruby extends VersionManager {
     return {
       env: { ...process.env, ...rubyEnv },
       yjit,
-      version: versionInfo.version,
+      version,
     };
   }
 
   // Returns the full URI to the Ruby executable
   protected async findRubyUri(rubyVersion: RubyVersion): Promise<vscode.Uri> {
+    if (/\d+\.\d+\.\d+/.exec(rubyVersion.version)) {
+      return this.findRubyUriForCompleteVersion(rubyVersion);
+    }
+
+    return this.findRubyUriWithOmittedPatch(rubyVersion);
+  }
+
+  private async findRubyUriWithOmittedPatch(
+    rubyVersion: RubyVersion,
+  ): Promise<vscode.Uri> {
+    const possibleVersionNames = rubyVersion.engine
+      ? [`${rubyVersion.engine}-${rubyVersion.version}`, rubyVersion.version]
+      : [rubyVersion.version, `ruby-${rubyVersion.version}`];
+
+    for (const uri of this.rubyInstallationUris) {
+      const directories = (await vscode.workspace.fs.readDirectory(uri)).sort(
+        (left, right) => right[0].localeCompare(left[0]),
+      );
+
+      for (const versionName of possibleVersionNames) {
+        const targetDirectory = directories.find(([name]) =>
+          name.startsWith(versionName),
+        );
+
+        if (targetDirectory) {
+          return vscode.Uri.joinPath(uri, targetDirectory[0], "bin", "ruby");
+        }
+      }
+    }
+
+    throw new Error(
+      `Cannot find installation directory for Ruby version ${possibleVersionNames.join(" or ")}.
+       Searched in ${this.rubyInstallationUris.map((uri) => uri.fsPath).join(", ")}`,
+    );
+  }
+
+  private async findRubyUriForCompleteVersion(
+    rubyVersion: RubyVersion,
+  ): Promise<vscode.Uri> {
     // If an engine was specified in the .ruby-version file, we favor looking for that first and also try just the
     // version number. If no engine was specified, we first try just the version number and then we try using `ruby` as
     // the default engine
@@ -99,46 +138,51 @@ export class Chruby extends VersionManager {
   private async discoverRubyVersion(): Promise<RubyVersion> {
     let uri = this.bundleUri;
     const root = path.parse(uri.fsPath).root;
+    let version: string;
+    let rubyVersionUri: vscode.Uri;
 
     while (uri.fsPath !== root) {
       try {
-        const rubyVersionUri = vscode.Uri.joinPath(uri, ".ruby-version");
+        rubyVersionUri = vscode.Uri.joinPath(uri, ".ruby-version");
         const content = await vscode.workspace.fs.readFile(rubyVersionUri);
-        const version = content.toString().trim();
-
-        if (version === "") {
-          throw new Error(`Ruby version file ${rubyVersionUri} is empty`);
-        }
-
-        const match =
-          /((?<engine>[A-Za-z]+)-)?(?<version>\d\.\d\.\d(-[A-Za-z0-9]+)?)/.exec(
-            version,
-          );
-
-        if (!match?.groups) {
-          throw new Error(
-            `Ruby version file ${rubyVersionUri} contains invalid format. Expected (engine-)?version, got ${version}`,
-          );
-        }
-
-        this.outputChannel.info(
-          `Discovered Ruby version ${version} from ${rubyVersionUri.toString()}`,
-        );
-        return { engine: match.groups.engine, version: match.groups.version };
+        version = content.toString().trim();
       } catch (error: any) {
         // If the file doesn't exist, continue going up the directory tree
         uri = vscode.Uri.file(path.dirname(uri.fsPath));
         continue;
       }
+
+      if (version === "") {
+        throw new Error(`Ruby version file ${rubyVersionUri} is empty`);
+      }
+
+      const match =
+        /((?<engine>[A-Za-z]+)-)?(?<version>\d+\.\d+(\.\d+)?(-[A-Za-z0-9]+)?)/.exec(
+          version,
+        );
+
+      if (!match?.groups) {
+        throw new Error(
+          `Ruby version file ${rubyVersionUri} contains invalid format. Expected (engine-)?version, got ${version}`,
+        );
+      }
+
+      this.outputChannel.info(
+        `Discovered Ruby version ${version} from ${rubyVersionUri.toString()}`,
+      );
+      return { engine: match.groups.engine, version: match.groups.version };
     }
 
     throw new Error("No .ruby-version file was found");
   }
 
   // Run the activation script using the Ruby installation we found so that we can discover gem paths
-  private async runActivationScript(
-    rubyExecutableUri: vscode.Uri,
-  ): Promise<{ defaultGems: string; gemHome: string; yjit: boolean }> {
+  private async runActivationScript(rubyExecutableUri: vscode.Uri): Promise<{
+    defaultGems: string;
+    gemHome: string;
+    yjit: boolean;
+    version: string;
+  }> {
     // Typically, GEM_HOME points to $HOME/.gem/ruby/version_without_patch. For example, for Ruby 3.2.2, it would be
     // $HOME/.gem/ruby/3.2.0. However, chruby overrides GEM_HOME to use the patch part of the version, resulting in
     // $HOME/.gem/ruby/3.2.2. In our activation script, we check if a directory using the patch exists and then prefer
@@ -155,7 +199,7 @@ export class Chruby extends VersionManager {
       "end",
       "newer_gem_home = File.join(File.dirname(user_dir), RUBY_VERSION)",
       "gems = (Dir.exist?(newer_gem_home) ? newer_gem_home : user_dir)",
-      "data = { defaultGems: Gem.default_dir, gemHome: gems, yjit: !!defined?(RubyVM::YJIT) }",
+      "data = { defaultGems: Gem.default_dir, gemHome: gems, yjit: !!defined?(RubyVM::YJIT), version: RUBY_VERSION }",
       "STDERR.print(JSON.dump(data))",
     ].join(";");
 

--- a/vscode/src/test/suite/ruby/chruby.test.ts
+++ b/vscode/src/test/suite/ruby/chruby.test.ts
@@ -135,6 +135,10 @@ suite("Chruby", () => {
     assert.match(env.GEM_PATH!, new RegExp(`lib/ruby/gems/${VERSION_REGEX}`));
     assert.strictEqual(version, RUBY_VERSION);
     assert.notStrictEqual(yjit, undefined);
+    fs.rmSync(path.join(rootPath, "opt", "rubies", `${RUBY_VERSION}-custom`), {
+      recursive: true,
+      force: true,
+    });
   });
 
   test("Considers Ruby as the default engine if missing", async () => {
@@ -180,5 +184,36 @@ suite("Chruby", () => {
     assert.match(env.GEM_PATH!, new RegExp(`lib/ruby/gems/${VERSION_REGEX}`));
     assert.strictEqual(version, RUBY_VERSION);
     assert.notStrictEqual(yjit, undefined);
+  });
+
+  test("Finds Ruby when .ruby-version omits patch", async () => {
+    fs.mkdirSync(
+      path.join(rootPath, "opt", "rubies", `${major}.${minor}.0`, "bin"),
+      {
+        recursive: true,
+      },
+    );
+
+    fs.writeFileSync(
+      path.join(workspacePath, ".ruby-version"),
+      `${major}.${minor}`,
+    );
+
+    const chruby = new Chruby(workspaceFolder, outputChannel);
+    chruby.rubyInstallationUris = [
+      vscode.Uri.file(path.join(rootPath, "opt", "rubies")),
+    ];
+
+    const { env, version, yjit } = await chruby.activate();
+
+    assert.match(env.GEM_PATH!, new RegExp(`ruby/${VERSION_REGEX}`));
+    assert.match(env.GEM_PATH!, new RegExp(`lib/ruby/gems/${VERSION_REGEX}`));
+    assert.strictEqual(version, RUBY_VERSION);
+    assert.notStrictEqual(yjit, undefined);
+
+    fs.rmSync(path.join(rootPath, "opt", "rubies", `${major}.${minor}.0`), {
+      recursive: true,
+      force: true,
+    });
   });
 });

